mailbox/backup/src/main/java/org/apache/james/mailbox/backup/zip/MailAccountZipIterator.java

```diff
@@ -0,0 +1,272 @@
+            .map(extraField -> ((StringExtraField) extraField).getValue())
+            .findFirst().flatMap(Function.identity());
+    }
+
+    private Optional<ZipEntryType> getEntryType(ZipEntry entry) throws ZipException {
+        ZipExtraField[] extraFields = ExtraFieldUtils.parse(entry.getExtra());
+        return Arrays.stream(extraFields)
+            .filter(field -> field.getHeaderId() == EntryTypeExtraField.ID_AQ)
+            .flatMap(extraField ->
+                OptionalUtils.toStream(((EntryTypeExtraField) extraField).getEnumValue()))
+            .findFirst();
+    }
+
+    private Optional<String> getMailBoxId(ZipEntry entry) throws ZipException {
+        return getStringExtraField(MailboxIdExtraField.ID_AM, entry);
+    }
+
+    private Optional<String> getMessageId(ZipEntry entry) throws ZipException {
+        return getStringExtraField(MessageIdExtraField.ID_AL, entry);
+    }
+
+    private Optional<Long> getSize(ZipEntry entry) throws ZipException {
+        return getLongExtraField(SizeExtraField.ID_AJ, entry);
+    }
+
+    private Optional<Date> getInternalDate(ZipEntry entry) throws ZipException {
+        ZipExtraField[] extraFields = ExtraFieldUtils.parse(entry.getExtra());
+        return Arrays.stream(extraFields)
+            .filter(field -> field.getHeaderId() == InternalDateExtraField.ID_AO)
+            .map(extraField -> ((InternalDateExtraField) extraField).getDateValue())
+            .findFirst()
+            .flatMap(Function.identity());
+    }
+
+    private Optional<Flags> getFlags(ZipEntry entry) throws ZipException {
+        return getStringExtraField(FlagsExtraField.ID_AP, entry).map(Flags::new);
+    }
+
+    private String getMailboxName(ZipEntry current) {
+        return current.getName().substring(0, current.getName().length() - 1);
+    }
+
+    private MailArchiveEntry fromMailboxEntry(ZipEntry current, Optional<ZipEntry> next) throws ZipException {
+        if (!next.isPresent() || !getEntryType(next.get()).equals(Optional.of(ZipEntryType.MAILBOX_ANNOTATION_DIR))) {
+            //no annotation directory or end of iterator => get Mailbox
+            currentMailBox = Optional.empty();
+            return new MailboxWithAnnotationsArchiveEntry(getMailboxName(current), getMailBoxId(current).get(), NO_ANNOTATION);
+        } else {
+            // annotation directory => save current mailbox and fetch next entry
+            currentMailBox = Optional.of(new MailboxWithAnnotationsArchiveEntry(getMailboxName(current), getMailBoxId(current).get(), NO_ANNOTATION));
+            return next();
+        }
+    }
+
+    private MailArchiveEntry fromMailboxAnnotationEntry(ZipEntry current, Optional<ZipEntry> next) throws ZipException {
+        MailboxAnnotation annotation = MailboxAnnotation.newInstance(getAnnotationKey(currentMailBox.get(), current), getAnnotationContent().orElse(""));
+        MailboxWithAnnotationsArchiveEntry mailboxUpdated = currentMailBox.get().appendAnnotation(annotation);
```

Comment:
Ok, this is the heart of all this uneeded complexity!

IMO we should find a way to retrieve the mailbox this annotation directly FROM the zip entry without needing to rely on the order of items.

If we are  able to revieve the mailbox, then we just:
 - Build an iterator of ZipEntry
 - Parse the ZipEntry to the corresponding type (Mailbox/Message/Annotations)
 - And this iterator will not need to track context + Parsing the ZipEntry can be delegated to another component

It's easy: the mailbox ID of the annotation is in the path you ignore....

Reply:
I don't think so. In my understanding the zip entry is mostly useless,  to read the content of one, i must be at the good position in the zipinputstream. The content is read from the zipinputstream not from an entry. At the beginning i hopped it will be easy as you just said but it proved otherwise.